### PR TITLE
hook create key route in ws-proxy

### PR DESCRIPTION
**Pull request by iQQBot**
## Description

hook create key route in ws-proxy, this provides backward compatibility for the older version of Jetbrains Gateway.

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at e1310c3</samp>

Add a new feature for generating SSH keys from the IDE. The `routes.go` file defines a new route and function that use ed25519 keys and return them in JSON format.

</details>

## Related Issue(s)

Fixes #

## How to test

1. start a worksapce
2. add `_supervisor/v1/ssh_keys/create` to the end of workspace URL, and access it
3. it should be output a privateKey(re-generate everytime), a hostKey(should be match one of ssh gateway host key, check by `_ssh/host_keys` endpoint), and a username (always gitpod)
4. start a workspace with jetbrains IDE, and use old gateway plugin (i.e. 18987) it should be success

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - pd-create-key-hook</li>
	<li><b>🔗 URL</b> - <a href="https://pd-create-key-hook.preview.gitpod-dev.com/workspaces" target="_blank">pd-create-key-hook.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - pd-create-key-hook-gha.21138</li>
	<li><b>🗒️ Logs</b> - <a href="https://console.cloud.google.com/logs/query;query=jsonPayload.kubernetes.host%3D%22preview-pd-create-key-hook%22%0A%0A--%20Filter%20on%20service:%0A--%20jsonPayload.serviceContext.service%3D%22ws-manager-mk2%22%0A;duration=P1D?project=gitpod-core-dev" target="_blank">GCP Logs Explorer</a></li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [x] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [x] /werft preemptible
      Saves cost. Untick this only if you're really sure you need a non-preemtible machine.
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

**Comment by kylos101** [comment]
Building preview env

**Comment by jeanp413** [comment]
@kylos101 I'm testing it right now with @iQQBot

**Comment by kylos101** [comment]
> @kylos101 I'm testing it right now with @iQQBot

👍 added you as an approver, @jeanp413

File: components/ws-proxy/pkg/proxy/routes.go [hunk 1]
```diff
@@ -179,6 +185,58 @@ func (ir *ideRoutes) HandleSSHHostKeyRoute(route *mux.Route, hostKeyList []ssh.S
...
+			log.WithError(err).Error("failed to generate key")
+			return
+		}
+		block, err := ssh.MarshalPrivateKey(pvk, "")
+		if err != nil {
+			log.WithError(err).Error("failed to marshal key")
+			return
+		}
+		resp.Privatekey = string(pem.EncodeToMemory(block))
+		resp.UserName = "gitpod"
```
[the remark this answers is not in the document]
Comment (kylos101) [reply components/ws-proxy/pkg/proxy/routes.go:216]: I see, so old way was constrained to this user, whereas new way is not.

File: components/ws-proxy/pkg/proxy/routes.go [hunk 2]
```diff
@@ -179,6 +185,58 @@ func (ir *ideRoutes) HandleSSHHostKeyRoute(route *mux.Route, hostKeyList []ssh.S
...
+
+		var hostKey ssh.Signer
+		for _, hk := range hostKeyList {
+			if hk.PublicKey().Type() != ssh.KeyAlgoRSA {
+				hostKey = hk
+				break
+			}
+			if hostKey == nil {
+				hostKey = hk
```
[the remark this answers is not in the document]
Comment (kylos101) [reply components/ws-proxy/pkg/proxy/routes.go:225]: I see. That makes sense, thank you! So this is a fallback, if hostkey is still nil.

**Review by jeanp413 — APPROVED** [review]
Tested with @iQQBot, everything works 👍

**Review by kylos101 — APPROVED** [review]

**Pull request merged**